### PR TITLE
fix default placeholder race condition on amp-ad

### PR DESCRIPTION
**Pull request by erwinmombay**

### CODE REVIEW

**Comment by erwinmombay** [comment]
Fixes #854

File: builtins/amp-ad.js [hunk 1]
```diff
@@ -243,21 +244,24 @@ export function installAd(win) {
...
         });
       }
       return loadPromise(this.iframe_);
     }
 
+    noContentHandler_() {
+      if (this.isDefaultFallback_) {
+        this.setDefaultFallback_();
+        this.element.appendChild(this.fallback_);
+        this.element.removeChild(this.iframe_);
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: Should we remove iframe_ in either case?

Comment (erwinmombay) [inline builtins/amp-ad.js]: good point. we should

File: builtins/amp-ad.js [hunk 2]
```diff
@@ -265,12 +269,11 @@ export function installAd(win) {
      * @private
      * visibleForTesting
      */
-    setDefaultPlaceholder_() {
+    setDefaultFallback_() {
       var a = document.createElement('a');
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: `const a`?

File: src/3p-frame.js [hunk 3]
```diff
@@ -102,16 +102,19 @@ export function getIframe(parentWindow, element, opt_type) {
...
  * @param {string} typeOfMessage
  * @param {function(!Object)} callback Called when a message of this type
  *     arrives for this iframe.
+ * @return {function()}
  */
 export function listen(iframe, typeOfMessage, callback) {
   var win = iframe.ownerDocument.defaultView;
   var origin = parseUrl(getBootstrapBaseUrl(win)).origin;
-  win.addEventListener('message', function(event) {
+  let listener = function(event) {
```
Comment (dvoytenko) [inline src/3p-frame.js]: `const`

File: builtins/amp-ad.js (hunk 2 above)
Comment (erwinmombay) [inline builtins/amp-ad.js]: done.

File: src/3p-frame.js [hunk 4]
```diff
@@ -102,16 +102,19 @@ export function getIframe(parentWindow, element, opt_type) {
...
+ * Allows listening for message from the iframe. Returns an unlisten
+ * function to remove the listener.
+ *
  * @param {!Element} iframe
  * @param {string} typeOfMessage
  * @param {function(!Object)} callback Called when a message of this type
  *     arrives for this iframe.
+ * @return {function()}
  */
 export function listen(iframe, typeOfMessage, callback) {
```
Comment (dvoytenko) [inline src/3p-frame.js]: Since we are here, we may as well create `listenOnce` similar to the ones in `event-helper.js`?

Comment (erwinmombay) [inline src/3p-frame.js]: yes, i thought about that. wanted to do it in a separate PR if thats ok with you. since we want to do a release once this bug is fixed.

https://github.com/ampproject/amphtml/issues/861 (create listenOnce helper)

Comment (dvoytenko) [inline src/3p-frame.js]: If it's just one case - we can skip it too. Not a big deal.

File: src/3p-frame.js [hunk 5]
```diff
@@ -102,16 +102,19 @@ export function getIframe(parentWindow, element, opt_type) {
...
 /**
- * Allows listening for message from the iframe.
+ * Allows listening for message from the iframe. Returns an unlisten
+ * function to remove the listener.
+ *
  * @param {!Element} iframe
  * @param {string} typeOfMessage
  * @param {function(!Object)} callback Called when a message of this type
  *     arrives for this iframe.
+ * @return {function()}
```
Comment (dvoytenko) [inline src/3p-frame.js]: It should be:
```
@return {!Unlisten}
```

File: builtins/amp-ad.js [hunk 6]
```diff
@@ -243,21 +244,24 @@ export function installAd(win) {
...
+        const unlisten = listen(this.iframe_, 'no-content', () => {
+          this.noContentHandler_();
+          unlisten();
         });
       }
       return loadPromise(this.iframe_);
     }
 
+    noContentHandler_() {
+      if (this.isDefaultFallback_) {
```
Comment (dvoytenko) [inline builtins/amp-ad.js]: Let's add a TODO/bug# to check on the performance of this thing. I don't like this relayout.

**Comment by dvoytenko** [comment]
LGTM with a couple of comments.

File: builtins/amp-ad.js (hunk 6 above)
Comment (erwinmombay) [inline builtins/amp-ad.js]: done. added TODO to check `append` as well as `display: block` on fallback.

File: src/3p-frame.js (hunk 5 above)
Comment (erwinmombay) [inline src/3p-frame.js]: done.

File: src/3p-frame.js (hunk 3 above)
Comment (erwinmombay) [inline src/3p-frame.js]: done.

File: builtins/amp-ad.js (hunk 1 above)
Comment (erwinmombay) [inline builtins/amp-ad.js]: done.
